packages/e2e/axe-selenium/src/main/java/org/talend/axeselenium/Reporter.java

```diff
@@ -0,0 +1,127 @@
+package org.talend.axeselenium;
+
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+import java.io.*;
+
+public class Reporter {
+    private static final String lineSeparator = System.getProperty("line.separator");
+
+    /**
+     * Build a formatted string violation report
+     * @param violations JSONArray of violations
+     * @return readable report of accessibility violations found
+     */
+    public String report(final JSONArray violations) {
+        final StringBuilder sb = new StringBuilder();
+        sb
+                .append("Found ")
+                .append(violations.length())
+                .append(" accessibility violations:");
+
+        for (int i = 0; i < violations.length(); i++) {
+            final JSONObject violation = violations.getJSONObject(i);
+            sb
+                    .append(lineSeparator)
+                    .append(i + 1)
+                    .append(") ")
+                    .append(violation.getString("help"));
+
+            if (violation.has("helpUrl")) {
+                sb
+                        .append(": ")
+                        .append(violation.getString("helpUrl"));
+            }
+
+            final JSONArray nodes = violation.getJSONArray("nodes");
+
+            for (int j = 0; j < nodes.length(); j++) {
+                final JSONObject node = nodes.getJSONObject(j);
+                sb
+                        .append(lineSeparator)
+                        .append("  ")
+                        .append(getOrdinal(j + 1))
+                        .append(") ")
+                        .append(node.getJSONArray("target"))
+                        .append(lineSeparator);
+
+                final JSONArray all = node.getJSONArray("all");
+                final JSONArray none = node.getJSONArray("none");
+
+                for (int k = 0; k < none.length(); k++) {
+                    all.put(none.getJSONObject(k));
+                }
+
+                appendFixes(sb, all, "Fix all of the following:");
+                appendFixes(sb, node.getJSONArray("any"), "Fix any of the following:");
+            }
+        }
+
+        return sb.toString();
+    }
```

Comment:
Isn't it possible to use `java.text.MessageFormat` or another template library to properly build that report?
Understanding the end result by looking at each line is impossible.

Reply:
Replace it with Formatter, bound to the StringBuilder, but it doesn't improve readability :/